Decide unboxing of let-bound expressions based on their Cmm translation

alainfrisch opened the pull request:
Previously, for a Clambda let-binding `let x = e1 in e2`, Cmmgen would look at the e1 Clambda expression to decide whether it is worth/possible to unbox the binding.  This requires to "predict" at the Clambda level if an expression produces a boxed number, while this is really a property of the Cmm code.  The analysis tried to replicate the behavior of the Clamdba->Cmm translation, but missed some cases such as `z` in:

````ocaml
let f x y =
  let z =
    let a = x +. y in
    let b = x *. y in
    if a <= b then
      a
    else
      b
  in
  z +. 1.
````

(It turns out that, recursively,`a`  and `b` will be unboxed, so accessing them will require a re-boxing, but when analyzing the `let z`, we still don't know that.)

The PR proposes to translate the bound Clambda expression to Cmm and analyze the result to decide if the binding should be unboxed.  This avoids a duplication of logic and fixes the code above.

Side-note: it is a bit unsatisfactory to reverse engineer boxing (but this was already the case before), and one might want to discuss keeping a symbolic representations of the boxing/unboxing operations in Cmm and lowering them later (as a simple Cmm->Cmm rewriting pass, or during the selection pass).

alainfrisch committed:
Decide unboxing of let-bound expressions based on their Cmm translation

alainfrisch commented:
I've also added a version of #2162 (unbox across static handlers), but now using the same unboxing strategy as for let bindings.

cc @chambart

alainfrisch commented:
@kayceesrk : would you be as kind as to add this PR to the benchmarking on http://ocamllabs.io/multicore/?

alainfrisch committed:
Changelog

gasche commented:
This is no criticism of @alainfrisch, but I find it mildly inappropriate to use @kayceesrk's time for non-multicore-related benchmarks because the compiler team uses no easily-deployed compiler benchmark tool. Is there an easy way to use a clone of the ocamllabs.io/multicore system (or some other, but maintained by someone) that gives what-to-benchmark control to compiler developers, so that @alainfrisch could directly start the measurements he was requesting above?

alainfrisch commented:
@gasche cf https://github.com/ocaml/ocaml/pull/2143#issuecomment-441009052

alainfrisch commented:
Mandatory stupid micro-benchmark:

````ocaml
let foo () =
  let r = ref 0. in
  let f x =
    r := !r +. x;
    if !r >= 1000. then r:= !r -. 1000.
  in
  for i = 1 to 100000 do
    match i with
    | 0 -> f 0.
    | 1 -> f 1.
    | n ->
        let a, b =
          if n > 100 then float n, float (n + 1)
          else float (n + 1), 42.
        in
        if n > 1000 then f (a *. b)
  done;
  !r


let () =
  for i = 1 to 1000 do
    ignore (foo ())
  done;
````

version | timing
--------|---------
OCaml 4.07 | 1.2s
Trunk (after #2143) |  0.40s
This PR | 0.25s

With this PR, the main function `foo` does not allocate at all (except for boxing its result).

alainfrisch commented:
@chambart @mshinwell do you think you could give a look to this PR?

lthls commented:
@gasche The [flambda benchmarks](http://bench.flambda.ocamlpro.com/) could be used instead (its infrastructure was actually built upon and improved for the multicore benchmarks), and allowing compiler developers to add/remove benchmarks would be as easy as giving them commit/merge rights to OCamlPro/ocamlbench-repo. Alternatively, anybody can submit a pull request to add a branch to test (there are several PRs that have been merged doing just this, for inspiration).

Although benchmarking the official compiler against proposed changes is already the purpose of the flambda benchmarks server, if someone wants to host another benchmarking server more clearly managed by the official OCaml team, we would be willing to help with setting up the infrastructure too.

gasche commented:
@lthls personally I would be very happy to use the flambda benchmarks. Where is the short documentation on how to add a new ocaml/ocaml PR there, with pointers to those pull requests you mention for inspiration?

("I would be happy to use" is more or less a figure of speech here: I don't personally submit many optimizations-related PRs, but others could be interested if the bar of entry wasn't too high. Clearly @alainfrisch doesn't think of your tool as something he could use easily, which suggests that something is slightly off right now. Thanks for replying here, which is a step to fix this!)

gasche commented:
Maybe we could add this documentation (on how to submit a PR to the flambda benchmarks) to our [CONTRIBUTING.md](https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md), or at least add a pointer to some documentation somewhere.

lthls commented:
I'm working on a pull request to add some information to CONTRIBUTING.md, but it looks like there are a few issues at the moment with our scripts (the version of trunk being benchmarked points to a commit from July at the moment, which is not intended), so I'll wait until we've fixed these issues before publishing the PR.

alainfrisch commented:
@chambart @mshinwell : do you think you might be able to review this PR?

lthls commented on asmcomp/cmm.ml:
```diff
@@ -220,3 +220,89 @@ let ccatch (i, ids, e1, e2)=
+      f body;
+      true
+  | Ctrywith(e1, _id, e2) ->
+      f e1;
```
While correct in the case of unboxing, keep in mind that `e1` here is not considered in tail position for the purpose of tailcall optimization. It might be worth mentioning in the `.mli` to prevent reuse in the wrong context.

lthls commented on asmcomp/cmm.ml:
```diff
@@ -220,3 +220,89 @@ let ccatch (i, ids, e1, e2)=
+      let map_h (n, ids, handler) = (n, ids, f handler) in
+      Ccatch(rec_flag, List.map map_h handlers, f body)
+  | Ctrywith(e1, id, e2) ->
+      Ctrywith(map_tail f e1, id, map_tail f e2)
```
Same as above

lthls reviewed (commented):
I noticed a few things that could be improved:

- There is a slight difference in the handling of `raise`: it used to be classified as `No_return` in `is_unboxed_number`, whereas now it falls into the default `No_unboxing` case of `is_unboxed_number_cmm`. I don't know how much of a difference it would make in practice, but it should be easy to make a special case for `Cop(Craise)` anyway.
Here is a micro-benchmark showing a difference:

```ocaml
let improbable () = Sys.opaque_identity false

let id x = x [@@inline]

let () =
  let x0 = Gc.minor_words () in
  let r = ref 0. in
  for i = 0 to 99999 do
    r := id (if improbable () then raise Exit else !r +. 1.)
  done;
  let x1 = Gc.minor_words () in
  Printf.printf "Minor words: %f\n%!" (x1 -. x0)
```

- There is a small issue with assuming the body of a `trywith` block to be in tail position, but this is a matter of properly documenting the behaviour, not a bug in the code (see inline comment).

I didn't notice any other problem with the patch, but I haven't reviewed it in depth.

alainfrisch commented:
> There is a slight difference in the handling of raise

Thanks.  I've fixed that by extending the two new `*_tail` iterators in Cmm to let them know that `Craise` never return a value.

alainfrisch commented:
@lthls : thanks for the review, and sorry for the slow reaction.  Now that you have Approver power, can you formally approve this PR if you are happy with the latest state?

I've rebased and fixed a number of conflicts, in particular with #2308.

mshinwell commented:
@alainfrisch Could we _please_ hold off merging this until the debugging changes to `Cmm` and `Cmmgen` have settled down?  I've had to write a new version of #2294, which includes changes to the `Cmm` language and `Cmmgen` again, and there is some urgency to get this resolved due to blocking many other patches.

alainfrisch commented:
Looking quickly at changes to Cmm/Cmmgen in #2294, there should be no (serious) conflict with this one.   I'm not strongly opposed to waiting a bit more for this PR, but it gives an immediate improvement and it has been ready for four months, so I'm not sure that adding it to the list of things blocked by #2294 is the best approach, one should rather try to decouple issues and merge them when they are ready.

lthls commented on asmcomp/cmm.mli:
```diff
@@ -223,12 +223,18 @@ val reset : unit -> unit
+      of "tail" sub-expression used here does not match the one used
+      to trigger tail calls; in particular, try...with handlers are
+      considered to be in tail position (because their result
+      become the final result for the xpression).
```
Typo: expression

alainfrisch replied:
Thanks, fixed.

mshinwell commented:
@alainfrisch You're probably not looking at the revised version of the code for #2294 though, which I think is going to conflict.  I'm hoping it won't take too long to get this new version of #2294 reviewed and merged, and I'm certainly happy to help fix the conflicts in this one after that.  (Edit: so we can make sure this gets in for 4.09 too.)

gasche commented on asmcomp/cmm.ml:
```diff
@@ -222,3 +222,92 @@ let ccatch (i, ids, e1, e2, dbg) =
+  | Cswitch(e, tbl, el, dbg') ->
+      Cswitch(e, tbl, Array.map (fun (e, dbg) -> map_tail f e, dbg) el, dbg')
+  | Ccatch(rec_flag, handlers, body) ->
+      let map_h (n, ids, handler, dbg) = (n, ids, f handler, dbg) in
```
Why is this `f handler` and not `map_tail f handler`?

alainfrisch replied:
Bad merge conflict resolution...; thanks, fixed (also on the call for `body`).

lpw25 commented:
Is #8584 relevant here? :smile:

mshinwell commented:
@alainfrisch A couple of comments:
- It doesn't look like this has received a thorough review yet -- @lthls said he hadn't looked in detail.  I'm sorry I haven't had time to look at it yet either.
- Do you have some benchmark results on large-scale applications?
- It seems like there's some scope for mistakes here since this touches a few tricky things (e.g. handling of 64-bit integers as pairs on 32-bit platforms); do you have any thoughts as to how we can gain confidence this is correct?

One thing I'm particularly interested in is whether this work forms a stepping-stone (and if so, how much of one) towards a potentially independent unboxing pass on Cmm code.  Or alternatively, if it couldn't be a fully independent pass, whether it could be incorporated into the "library" part of #1963. 
 The reason I'm interested is that, if this were the case, we could potentially use it in a first version of Flambda 2.0 (which is aiming to replicate approximately the functionality of Closure + Cmmgen) instead of having to implement unboxing translations within Flambda at that point.  (We plan for Flambda 2.0 to be able to do all of the unboxing itself in the medium term.)

alainfrisch commented:
> Is #8584 relevant here? 😄

Yes, of course, absolutely.  Hence https://github.com/ocaml/ocaml/pull/2165#issuecomment-442557042

alainfrisch commented on asmcomp/cmmgen.ml:
```diff
@@ -1107,49 +1118,60 @@ let alloc_matches_boxed_int bi ~hdr ~ops =
+    else
+      Cop(
+        Cload((if bi = Pint32 then Thirtytwo_signed else Word_int),
+              Immutable),
```
Note the change from `Mutable` to `Immutable`, unrelated to this PR, but correct AFAICT.

lpw25 commented:
Ah, I'd missed that one.

alainfrisch commented:
> It seems like there's some scope for mistakes here since this touches a few tricky things (e.g. handling of 64-bit integers as pairs on 32-bit platforms); do you have any thoughts as to how we can gain confidence this is correct?

Are you referring to https://github.com/ocaml/ocaml/blob/26f7bd9fa247a62a6678779a59ff214c18722c46/asmcomp/cmmgen.ml#L1161-L1168 ?  I think that all other stuff related to 64-bit integers on 32-bit platforms is inherited from the current code.

> Do you have some benchmark results on large-scale applications?

Nope.  But I know that the code sample in the description and the micro-benchmark are not too far away from code that can appear in real-world numerical code.  Also, this is a good companion to the new compilation scheme for local functions, which can introduce more static handlers than inlining (and this PR would avoid adding more boxing in those cases).

> One thing I'm particularly interested in is whether this work forms a stepping-stone (and if so, how much of one) towards a potentially independent unboxing pass on Cmm code.

I'd say the opposite actually, since it brings unboxing closer to the Cmm world than to the translation from Clambda to Cmm.   I haven't looked at #1963 closely enough to know whether it would fit in its library part, but very possibly so.  But please don't make that big refactoring yet another prerequisite for the current PR.

mshinwell commented:
I don't understand your last paragraph.  If the unboxing is brought closer to the Cmm world -- and in particular makes it independent of Clambda -- then that would seem to form the stepping-stone that I'm talking about.

alainfrisch commented:
Ah sorry, I thought a stepping stone was something bad (an obstacle which makes you fall); I realize it's something good.  Yes, I confirm that the direction taken here should simplify doing things more at the Cmm level.

mshinwell commented:
@alainfrisch How hard would it be to split this PR into two, one based on the other, so we can read the unbox-handlers part separately?

alainfrisch commented:
> How hard would it be to split this PR into two, one based on the other, so we can read the unbox-handlers part separately?

You should get a clear picture of the two main parts (cmm-based unboxing and unboxing across static-handlers) if you look at the individual commits, I think.

mshinwell commented:
I was just thinking about trying to get some benchmarking results for this and it seems important to consider the changes independently (presumably the Clambda/Cmm change shouldn't change the generated code, whereas the other one will?).  The most clear way of doing this, especially given that there will likely be further commits dealing with review comments, would be to have two separate PRs rather than selecting subsets of commits from an existing one.  I'm not a fan of excessive splitting up of PRs because our workflow is such that this can occupy a disproportionate amount of time, but in this case it sounds like it should be ok.

alainfrisch commented:
I've rewritten the history to have just two meaningful and standalone commits for the two parts.  Do you really prefer two PRs?

mshinwell commented:
I think so -- that's our standard practice, no?  These features I hope are orthogonal to each other...

I'm ready to start reviewing pretty soon, by the way.

alainfrisch commented:
> These features I hope are orthogonal to each other...

The second commit is based on the first one, though.

I've kept only the first part in the PR, and move thed second part in #8735 (which also contains commits from this PR).

lthls commented on asmcomp/cmm.ml:
```diff
@@ -210,3 +210,54 @@ let ccatch (i, ids, e1, e2, dbg) =
+      true
+  | Cexit _ | Cop (Craise _, _, _) ->
+      true
+  | _ ->
```
You're missing the `Cphantom_let` case. Matching explicitly on all the remaining cases (instead of `_`) would be more future-proof, although perhaps a bit heavy.

alainfrisch replied:
Well spotted, thanks.  Will switch to explicit listing remaining cases indeed.

lthls commented on asmcomp/cmm.ml:
```diff
@@ -210,3 +210,54 @@ let ccatch (i, ids, e1, e2, dbg) =
+      Ctrywith(map_tail f e1, id, map_tail f e2, dbg)
+  | Cexit _ | Cop (Craise _, _, _) as cmm ->
+      cmm
+  | c ->
```
Same as above

lthls commented on asmcomp/cmmgen.ml:
```diff
@@ -64,6 +64,13 @@ let add_unboxed_id id unboxed_id bn env =
     unboxed_ids = V.add id (unboxed_id, bn) env.unboxed_ids;
   }
 
+let constants_tbl = Hashtbl.create 16
```
I feel like this could be part of the environment passed during translation instead of a global variable. As far as I can tell it is only modified at the very beginning of the compilation unit's translation, so there would be no particular need for this structure to be mutable either (except performance, possibly).
Otherwise, there is a recent file called `Cmmgen_state` that would be a nice place to put this variable.

alainfrisch replied:
I agree on principle, but I'll go with the suggestion to use Cmmgen_state.  It would make sense to move all of it into an environment passed through Cmmgen (there is already an Cmmgen.env, but the extra plumbing is still a bit invasive), but this can be done later.

alainfrisch commented:
Gentle ping to @mshinwell or whoever else (@lthls ?) interested in reviewing this PR and its companion #8735.

lthls reviewed (approved):
Globally nice changes.
In addition to a few specific remarks, I'll note that this work makes it easier to separate unboxing from translation. I don't think it's worth pushing it further and make unboxing a separate pass at the moment, but it's a good thing in my opinion.

(I had this review done a while ago, but I forgot I didn't submit it. Sorry for the delay.)

alainfrisch committed:
Take reviewer comments into account

alainfrisch commented:
Thanks @lthls for your positive review!  I've integrated your comments (last commit).  If you are ok, can you merge?

alainfrisch committed:
Merge remote-tracking branch 'origin/trunk' into unbox_cmm

lthls commented:
I'm going to merge this soon. On the kind of merge: I tend to prefer squash and merge, but I've seen that regular merges are quite common. Is there a global policy or are we free to choose based on personal taste ?

alainfrisch commented:
Thanks!

I'm not aware of a global policy but I also usually prefer squash-and-merge, and for this specific case, this seems to be right choice (there is only one "real" commit).

gasche commented:
It's important for me that there is something that looks like a "merge commit" in the result, notably that contains the number of the Github PR for later reference (merge commits from github always have that). I think that squash-and-merge is fine in that regard -- rebase-and-merge is the one that should be avoided.

lthls commented:
Squashed and merged, then. Thanks @alainfrisch for the PR.

alainfrisch commented:
Great, thanks to you for the review!

If you still have some spare cycles, #8735 is waiting for you 😄   It is a spin-off from this PR (see the benchmark above).
